### PR TITLE
feat:create class to monitor pi system

### PR DESCRIPTION
**Pull request by fe51**
This feature aims to monitor some metrics from raspberry pi system.

 In order to design a suitable system, I think some metrics could be recorded. 
Then, after test phases, especially if the system is not connected to the internet, the output can be analyzed to understand how to improve the system or understand why it crashed.
As an example, in order to know if fans, heatsink are needed, cpu temperature can be recorded; Or, some performance about cpu usage or memory available could help to select the best raspberry pi model fitting our needs.

Output of this feature is a csv, an example is provided in the following zip file [pi_perf.csv.zip](https://github.com/pyronear/PyroNear/files/5010251/pi_perf.csv.zip) and below

|datetime      |cpu_temperature_C|mem_available_GB|cpu_usage_percent|
|--------------|-----------------|----------------|-----------------|
|01/08/20 12:26|57.94            |3.00            |0.0              |
|01/08/20 12:26|56.97            |3.00            |0.3              |
|01/08/20 12:26|57.45            |3.00            |0.2              |
|01/08/20 12:26|56.97            |3.00            |0.2              |

This is a first draft, structure/approach can be discussed and new metrics suggestion are welcomed.

Also, I think I might need to update requirements file.

### CODE REVIEW

File: raspberryPi/monitorPi.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+
+class MonitorPi:
+    """This class aims to monitor some metrics from Raspberry Pi system.
+
+    Example
+    --------
+    recordlogs = MonitorPi("/home/pi/Desktop/")
+    recordlogs.record(5) # record metrics every 5 seconds
+
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: There is a trailing whitespace here, would you mind removing it please?

File: raspberryPi/monitorPi.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+        # cpu temperature
+        self.cpu_temp = CPUTemperature()
+
+        # path
+        self.monitoringFolder = monitoringFolder
+        self.logFile = logFile
+
+    def record(self, timeStep):
+
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: Same here

File: raspberryPi/monitorPi.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        # path
+        self.monitoringFolder = monitoringFolder
+        self.logFile = logFile
+
+    def record(self, timeStep):
+            
+        with open(self.monitoringFolder + self.logFile, "a") as log:
+            log.write("datetime;cpu_temperature_C;mem_available_GB;cpu_usage_percent\n")
+
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: Same here

File: raspberryPi/monitorPi.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+        self.monitoringFolder = monitoringFolder
+        self.logFile = logFile
+
+    def record(self, timeStep):
+            
+        with open(self.monitoringFolder + self.logFile, "a") as log:
+            log.write("datetime;cpu_temperature_C;mem_available_GB;cpu_usage_percent\n")
+                        
+            while True:
+                line = "{date};{t:.2f};{mem:.2f};{cpu_usage}\n".format(date=strftime("%Y-%m-%d %H:%M:%S"),t=self.cpu_temp.temperature,mem=psutil.virtual_memory().available / 1024**3,cpu_usage=psutil.cpu_percent())
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: Can you add whitespaces right after commas please?

Comment (frgfm) [reply raspberryPi/monitorPi.py]: The line is too long as well, perhaps break line between commas (no need for white spaces then)

File: raspberryPi/monitorPi.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: Whitespace to remove here

File: raspberryPi/monitorPi.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    def record(self, timeStep):
+            
+        with open(self.monitoringFolder + self.logFile, "a") as log:
+            log.write("datetime;cpu_temperature_C;mem_available_GB;cpu_usage_percent\n")
+                        
+            while True:
+                line = "{date};{t:.2f};{mem:.2f};{cpu_usage}\n".format(date=strftime("%Y-%m-%d %H:%M:%S"),t=self.cpu_temp.temperature,mem=psutil.virtual_memory().available / 1024**3,cpu_usage=psutil.cpu_percent())
+                    
+                log.write(line)
+
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: Whitespace to be removed

File: raspberryPi/monitorPi.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+                    
+                log.write(line)
+                
+                sleep(timeStep)
+
+
+if __name__ == "__main__":
+
+    recordlogs = MonitorPi("/home/pi/Desktop/")
+    recordlogs.record(30)
```
Comment (frgfm) [inline raspberryPi/monitorPi.py]: Could you add a blankline at the end of the file please?

**Review by frgfm — CHANGES_REQUESTED** [review]
Hey there, thanks for the PR!
Generally speaking, to check why some tests fail (the ones with red crosses), click on `Details` at their very right, it will take you to the code that was executed to run the test. In this case, only a few "style" changes to make!

Also, could you provide an example of usage and output for this feature please?

**Comment by fe51** [comment]
Hey @frgfm, thanks for the review !
I have updated the initial comment and style issues !

**Comment by frgfm** [comment]
Thanks for the edits!
Regarding the cpu usage, I don't remember well how `psutils` handles this, but is it memory usage or processing capacity usage?

Two quick things that might even improve the PR:
- adding the corresponding dependencies in the `requirements.txt`
- adding a unittest to check that we actually get the expected behaviour 
Happy to help if you need some :)

**Review by Akilditu — COMMENTED** [review]
Great idea, following our last talk it seems that the next step is to agree on where this PR should be merged : 

1. Having Rasp related codes and docs within /pyronear/PyroNear/
2. Create a dedicated repository like /pyronear/raspberryPi/

**Comment by MateoLostanlen** [comment]
@Akilditu and @fe51, I'm setuping the following repo https://github.com/pyronear/PyronearEngine. As discuss on the last metting this repo will regroup all work about deployment. This PR should be move there once it's ready. I let you know when it's done ;)

**Comment by fe51** [comment]
Sounds good to me !
Thank you @MateoLostanlen !

**Comment by frgfm** [comment]
@fe51 following up on our last discussions about repo separation, I would suggest moving the PR to the repo that will handle RaspberryPI utilities.

If so, it could be organized as follows:
- PyroNear: core package serving our trained models and available datasets
- PyronearEngine: on-device stack for RaspberryPI using PyroNear to serve models
- XYZ: we could also set up a web-app / interface to get demos on different devices

The main idea is to use PyroNear as core API block, that can be embedded on device (raspberry PI), in the browser (live web-app) or remain on a remote server (so that it could be called by a device / service).

In this scheme, this PR might be more useful in PyronearEngine :) If so, I would suggest to close this PR and open another on the repo. Let me know your thought on this!

**Comment by MateoLostanlen** [comment]
Hi @fe51, please can you close this PR and reopen it in https://github.com/pyronear/PyronearEngine ?

**Comment by fe51** [comment]
Hi @MateoLostanlen, I have just open a new PR in PyronearEngine : https://github.com/pyronear/PyronearEngine/pull/4

**Pull request closed without merging**